src/nix/path-info.cc

```diff
@@ -87,41 +96,98 @@ struct CmdPathInfo : StorePathsCommand, MixJSON
+
+                    clone["substitute"] = !infos.empty();
+
+                    auto jsonList(jsonList_.lock());
+                    jsonList->emplace_back(clone);
+                };
+
+                for (auto v : json) {
+                    pool.enqueue(std::bind(checkSubs, v));
+                }
+                pool.process();
+
+                auto jsonList(jsonList_.lock());
+                std::cout << jsonList->dump();
+            }
+            else std::cout << json.dump();
         }
 
         else {
+            ThreadPool pool(fileTransferSettings.httpConnections);
 
-            for (auto & storePath : storePaths) {
+            Sync<std::set<std::string>> results_;
+
+            std::function<void(StorePath)> printPath;
+            printPath = [&](const StorePath & storePath) {
                 auto info = store->queryPathInfo(storePath);
                 auto storePathS = store->printStorePath(info->path);
 
-                std::cout << storePathS;
+                if (showSubStatus) {
+                    SubstitutablePathInfos infos;
+                    store->querySubstitutablePathInfos({{storePath, std::nullopt}}, infos);
+
+                    if (!infos.empty()) {
+                        return;
+                    }
+                }
 
                 if (showSize || showClosureSize || showSigs)
-                    std::cout << std::string(std::max(0, (int) pathLen - (int) storePathS.size()), ' ');
+                    storePathS.append(std::max(0, (int) pathLen - (int) storePathS.size()), ' ');
 
                 if (showSize)
-                    printSize(info->narSize);
+                    storePathS.append(getSize(info->narSize));
 
                 if (showClosureSize)
-                    printSize(store->getClosureSize(info->path).first);
+                    storePathS.append(getSize(store->getClosureSize(info->path).first));
 
                 if (showSigs) {
-                    std::cout << '\t';
+                    storePathS.append(1,'\t');
                     Strings ss;
                     if (info->ultimate) ss.push_back("ultimate");
                     if (info->ca) ss.push_back("ca:" + renderContentAddress(*info->ca));
                     for (auto & sig : info->sigs) ss.push_back(sig);
-                    std::cout << concatStringsSep(" ", ss);
+                    storePathS.append(concatStringsSep(" ", ss));
                 }
 
-                std::cout << std::endl;
+                auto results(results_.lock());
+                results->emplace(storePathS);
+            };
+
+            for (auto & storePath : storePaths) {
+                pool.enqueue(std::bind(printPath, storePath));
             }
+            pool.process();
 
+            auto results(results_.lock());
+            for (auto & str : *results) {
+                std::cout << std::endl << str;
+            }
```

Comment:
`std::endl` means to emit a new line character *and* flush the terminal. But this code suggests that you mean to separate all strings by newlines and then flush only once at the end, so i suggest this:

```suggestion
            auto results(results_.lock());
            for (const auto & str : *results_.lock()) {
                std::cout << '\n' << str;
            }
            std::cout << std::endl;
```